src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECParameters.cs

```diff
@@ -62,5 +66,374 @@ public void Validate()
 
             Curve.Validate();
         }
+
+        internal static unsafe ECParameters FromECPrivateKey(ReadOnlySpan<byte> key, out int bytesRead)
+        {
+            fixed (byte* ptr = &MemoryMarshal.GetReference(key))
+            {
+                using (MemoryManager<byte> manager = new PointerMemoryManager<byte>(ptr, key.Length))
+                {
+                    ECPrivateKey parsedKey =
+                        AsnSerializer.Deserialize<ECPrivateKey>(manager.Memory, AsnEncodingRules.BER, out bytesRead);
+
+                    ECParameters ret;
+                    AlgorithmIdentifierAsn algId = default;
+                    FromECPrivateKey(parsedKey, algId, out ret);
+                    return ret;
+                }
+            }
+        }
+
+        internal static void FromECPrivateKey(
+            in ECPrivateKey key,
+            in AlgorithmIdentifierAsn algId,
+            out ECParameters ret)
+        {
+            ValidateParameters(key.Parameters, algId);
+
+            if (key.Version != 1)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            // Implementation limitation
+            if (key.PublicKey == null)
+            {
+                throw new CryptographicException(SR.Cryptography_NotValidPublicOrPrivateKey);
+            }
+
+            ReadOnlySpan<byte> publicKeyBytes = key.PublicKey.Value.Span;
+
+            if (publicKeyBytes.Length == 0)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            // Implementation limitation
+            // 04 (Uncompressed ECPoint) is almost always used.
+            if (publicKeyBytes[0] != 0x04)
+            {
+                throw new CryptographicException(SR.Cryptography_NotValidPublicOrPrivateKey);
+            }
+
+            // https://www.secg.org/sec1-v2.pdf, 2.3.4, #3 (M has length 2 * CEIL(log2(q)/8) + 1)
+            if (publicKeyBytes.Length != 2 * key.PrivateKey.Length + 1)
```

Comment:
is PrivateKey.Length already `log2(q)/8`?

Reply:
Yeah, that means "the size in bytes"